### PR TITLE
feat(spanner): support custom setting of timeout and retry

### LINKED ISSUE #7011: spanner: add support for custom timeout and retry options - Google::Gax::CallOptions
Currently, the client lib does not allow users to provide custom timeout and retry options that overrides the gapic default settings. 

`Google::Gax::CallOptions` is supported by the methods in spanner_client: 

https://github.com/googleapis/google-cloud-ruby/blob/1b0545aef072289a5d88a42054f9bed582423067/google-cloud-spanner/lib/google/cloud/spanner/v1/spanner_client.rb#L829

However, the methods in `service.rb` do not accept `Google::Gax::CallOptions`: 

https://github.com/googleapis/google-cloud-ruby/blob/1b0545aef072289a5d88a42054f9bed582423067/google-cloud-spanner/lib/google/cloud/spanner/service.rb#L304-L307

We can implement the following two features to allow overrides: 

- Method-level: update the methods in `service.rb` to support CallOptions. 
- Client-level: accept a hash/map (mapping a method name to a CallOptions) when initializing the service and update `default_options_from_session` to merge the custom CallOptions of a method. 

The precedence of order: 

gapic defaults < client-level < method-level

@jiren Do you have any thoughts on this? Thanks.

### PR DESCRIPTION
**Pull request by hengfengli**
Changes included: 

- Add `call_options` to allow setting custom timeout and retry policy at method-level. 
- Fix all related tests. 

An example: 

```
    timeout = 30
    retry_policy = {
      initial_delay: 0.25,
      max_delay:     32.0,
      multiplier:    1.3,
      retry_codes:   ["UNAVAILABLE"]
    }
    call_options = { timeout: timeout, retry_policy: retry_policy }

    results = client.execute_query "SELECT * FROM users", call_options: call_options
```

The following changes are backward incompatible, which we do not include them in this PR: 

- Add `call_options` to `upsert`, `insert`, `update`, `replace` methods in `lib/google/cloud/spanner/client.rb`. 

If we add the new parameter `call_options`, the following use case would fail: 

```ruby
db.upsert table_name, { id: id, timestamps: [Time.parse("2016-12-30 00:00:00Z"), Time.parse("2016-12-31 00:00:00Z"), Time.parse("2017-01-01 00:00:00Z")] }
```

This is because it treats the hash as a hash of function parameters instead of the first parameter. A fix would be to make it an array instead of a hash: 

```ruby
db.upsert table_name, [{ id: id, timestamps: [Time.parse("2016-12-30 00:00:00Z"), Time.parse("2016-12-31 00:00:00Z"), Time.parse("2017-01-01 00:00:00Z")] }]
```

But still, this requires users to update their code. To avoid this, we do not add `call_options` to the above methods. 

Fixes #7011

### CODE REVIEW

File: google-cloud-spanner/lib/google/cloud/spanner/batch_snapshot.rb [hunk 1]
```diff
@@ -173,6 +173,14 @@ def timestamp
         #   * `:optimizer_version` (String) The version of optimizer to use.
         #     Empty to use database default. "latest" to use the latest
         #     available optimizer version.
+        # @param [Hash] call_options A hash of values to specify the custom
+        #   call options, e.g., timeout, retries, etc. Call options are
+        #   optional. The following settings can be provided:
+        #
+        #   * `:timeout` (Numeric) A numeric value of custom timeout in seconds
+        #     that overrides the default setting.
+        #   * `:retry_policy` (Hash) A hash of values that overrides the default
```
Comment (jiren) [inline google-cloud-spanner/lib/google/cloud/spanner/batch_snapshot.rb:182]: Can you add attributes docs for retry_policy or ref to doc link.

i.e https://github.com/googleapis/google-cloud-ruby/blob/8fc158a134ea29a8435f6681b20f9e2600d3e0b7/google-cloud-spanner-v1/lib/google/cloud/spanner/v1/spanner/client.rb#L1761-L1767

File: google-cloud-spanner/lib/google/cloud/spanner/service.rb [hunk 2]
```diff
@@ -466,9 +559,18 @@ def lib_name_with_prefix
           value << " gccl"
         end
 
-        def default_options_from_session session_name
-          default_prefix = session_name.split("/sessions/").first
-          { metadata: { "google-cloud-resource-prefix" => default_prefix } }
+        def default_options session_name: nil, call_options: nil
```
Comment (jiren) [inline google-cloud-spanner/lib/google/cloud/spanner/service.rb:562]: Instead if unless and .nil? I think do something like this.

```ruby
 def default_options session_name: nil, call_options: nil
  opts = {}

  if session_name
    default_prefix = session_name.split("/sessions/").first
    opts[:metadata] = { "google-cloud-resource-prefix" => default_prefix }
  end

  if call_options
    opts[:timeout] = call_options[:timeout] if call_options[:timeout]
    opts[:retry_policy] = call_options[:retry_policy] if call_options[:retry_policy]
  end

  return opts unless opts.empty?
end
```

File: google-cloud-spanner/lib/google/cloud/spanner/batch_snapshot.rb (hunk 1 above)
Comment (hengfengli) [reply google-cloud-spanner/lib/google/cloud/spanner/batch_snapshot.rb:182]: Thanks for the suggestion. Fixed.

**Comment by jiren** [comment]
LGTM

**Pull request merged**